### PR TITLE
ontology annotation validation

### PR DESCRIPTION
**Pull request by erik-nieh**
## 🔗 Related Issues
https://github.com/voxel51/fiftyone/pull/7341

## 📋 What changes are proposed in this pull request?

Adds validation logic for verifying integrity of conditional attributes within a annotation ontology

This is not wired in anywhere yet, and this may be expanded (or contracted) later.

## 🧪 How is this patch tested? If it is not, please explain why.

Tested locally, unit tests. 

### Is this a user-facing change that should be mentioned in the release notes?

-   [x] No. You can skip the rest of this section.
-   [ ] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

add support for ontology objects to the SDK. 

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [x] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

## Summary by CodeRabbit

* **New Features**
  * Added validation for annotation ontologies: checks for duplicate attribute names, allowed condition operators, type/component compatibility, permitted override keys, cyclic dependency detection, and automatic validation when saving an ontology.
  * Expanded type-to-component mappings to improve compatibility checks and error reporting.

* **Tests**
  * Added tests covering validation failures and edge cases, including mutated conditions, unsupported types, component mismatches, invalid overrides, and cyclic dependencies.

### CODE REVIEW

File: tests/unittests/ontology_validation_tests.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            attributes=[
+                AttributeSpec(name="flag", type="bool", component="dropdown"),
+            ],
+        )
+        with self.assertRaises(ValueError):
+            validate_annotation_ontology(ao)
+
+
+class NoCyclesTests(unittest.TestCase):
+    def test_rejects_cycle_between_attributes(self):
```
Comment (benjaminpkane) [inline tests/unittests/ontology_validation_tests.py:73]: Awesome

File: fiftyone/core/ontology_validation.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+from fiftyone.core.ontology import AnnotationOntology
+
+
+_ALLOWED_THEN_KEYS = {"values", "component"}
+
+
+def validate_annotation_ontology(ontology: AnnotationOntology) -> None:
+    """Validates an :class:`AnnotationOntology`.
+
+    Runs every rule and aggregates all failures into a single error.
```
Comment (benjaminpkane) [inline fiftyone/core/ontology_validation.py]: Except field rules, yes? Can this function optionally take a label schema?

File: fiftyone/core/ontology_validation.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    # TODO: taxonomy-reference resolution is deferred until Phase 2 ships
+    # `Taxonomy` documents; validate that each `ontology.taxonomies` name
+    # resolves to a `Taxonomy` in the DB once the type exists.
+    errors: list[str] = [
+        *_validate_unique_attribute_names(ontology),
+        *_validate_when_operators(ontology),
+        *_validate_types(ontology),
+        *_validate_components(ontology),
+        *_validate_no_cycles(ontology),
+        *_validate_then_keys(ontology),
```
Comment (benjaminpkane) [inline fiftyone/core/ontology_validation.py]: Nice!

File: fiftyone/core/ontology_validation.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+    if duplicates:
+        return [f"duplicate attribute name(s): {sorted(duplicates)}"]
+    return []
+
+
+def _validate_when_operators(ontology: AnnotationOntology) -> list[str]:
+    """Safety net for ``When.operator`` — ``When.__post_init__`` rejects
+    invalid operators at construction, so this only catches values
+    mutated onto a constructed instance.
```
Comment (benjaminpkane) [inline fiftyone/core/ontology_validation.py:54]: Very detailed

File: fiftyone/core/ontology_validation.py (hunk 2 above)
Comment (erik-nieh) [reply fiftyone/core/ontology_validation.py]: I updated the docstring to be more specific than this is ontology - I think for now, it'd be best just to let it be specific. And I will add a call in the label schema validation itself to validate this. since we have to do checks to make sure that the applied ontology does exist at runtime anyways

File: fiftyone/core/annotation/constants.py [hunk 5]
```diff
@@ -88,6 +90,24 @@
...
+    DATETIME: DATE_DATETIME_COMPONENTS,
+    DICT: DICT_COMPONENTS,
+    FLOAT: FLOAT_INT_COMPONENTS,
+    FLOAT_LIST: FLOAT_INT_LIST_COMPONENTS,
+    ID: ID_COMPONENTS,
+    INT: FLOAT_INT_COMPONENTS,
+    INT_LIST: FLOAT_INT_LIST_COMPONENTS,
+    STR: STR_COMPONENTS,
+    STR_LIST: STR_LIST_COMPONENTS,
+}
```
Comment (benjaminpkane) [inline fiftyone/core/annotation/constants.py:111]: Helpful

**Review by benjaminpkane — APPROVED** [review]
